File: test/slug_test.dart
import 'package:diene_core_utils/diene_core_utils.dart';
import 'package:test/test.dart';

void main() {
  group('slugify', () {
    final Map<String, String> fixtures = <String, String>{
      'Hello World': 'hello-world',
      '  Trim Me  ': 'trim-me',
      'Multiple   Spaces': 'multiple-spaces',
      'Symbols!@#Here': 'symbols-here',
      'already-slugged': 'already-slugged',
      'mañana': 'manana',
      'résumé café': 'resume-cafe',
      '!!! ???': '',
    };

    for (final MapEntry<String, String> fixture in fixtures.entries) {
      test('normalizes ${fixture.key}', () {
        // Arrange
        final String input = fixture.key;

        // Act
        final String actual = slugify(input);

        // Assert
        expect(actual, fixture.value);
      });
    }
  });

  group('namespacedKey', () {
    test('joins normalized components without throwing', () {
      // Arrange
      const String namespace = 'Dart Core Utils';
      const String key = 'Sample Key';

      // Act
      final NamespacedKeyResult actual = namespacedKey(namespace, key);

      // Assert
      expect(actual, const ValidNamespacedKey('dart-core-utils:sample-key'));
      expect(actual.isValid, isTrue);
      expect(actual.hashCode, 'dart-core-utils:sample-key'.hashCode);
      expect(
        actual.match(valid: (String value) => value, invalid: (_) => 'bad'),
        'dart-core-utils:sample-key',
      );
    });

    test('returns namespace validation as data', () {
      // Arrange
      const String namespace = '!!!';

      // Act
      final NamespacedKeyResult actual = namespacedKey(namespace, 'key');

      // Assert
      expect(
        actual,
        const InvalidNamespacedKey(
          NamespacedKeyValidationError(
            field: NamespacedKeyField.namespace,
            reason: 'must not be empty',
          ),
        ),
      );
      expect(actual.isValid, isFalse);
      expect(
        actual.hashCode,
        const NamespacedKeyValidationError(
          field: NamespacedKeyField.namespace,
          reason: 'must not be empty',
        ).hashCode,
      );
      expect(
        actual.match(valid: (_) => 'bad', invalid: (error) => error.toString()),
        'namespace must not be empty',
      );
    });

    test('returns key validation as data', () {
      // Arrange
      const String key = '???';

      // Act
      final NamespacedKeyResult actual = namespacedKey('namespace', key);

      // Assert
      expect(
        actual,
        const InvalidNamespacedKey(
          NamespacedKeyValidationError(
            field: NamespacedKeyField.key,
            reason: 'must not be empty',
          ),
        ),
      );
    });
  });
}

File: tests/unit/slug.test.ts
import { describe, it } from 'bun:test';
import should from 'should';
import { NamespacedKeyValidationError, namespacedKey, slugify } from '../../src/slug';

describe('slugify', () => {
  it.each([
    { input: 'Hello World', expected: 'hello-world' },
    { input: '  Trim Me  ', expected: 'trim-me' },
    { input: 'Multiple   Spaces', expected: 'multiple-spaces' },
    { input: 'Symbols!@#Here', expected: 'symbols-here' },
    { input: 'already-slugged', expected: 'already-slugged' },
    { input: 'mañana', expected: 'manana' },
    { input: 'résumé café', expected: 'resume-cafe' },
    { input: '!!! ???', expected: '' },
  ])('should slugify "$input" to "$expected"', ({ input, expected }) => {
    // Arrange

    // Act
    const actual = slugify(input);

    // Assert
    should(actual).equal(expected);
  });
});

describe('namespacedKey', () => {
  it('should dogfood the published Result package for a valid key', async () => {
    // Arrange
    const namespace = 'Bun Core Utils';
    const key = 'Published Result';

    // Act
    const actual = namespacedKey(namespace, key);

    // Assert
    should(await actual.isOk()).equal(true);
    should(await actual.unwrap()).equal('bun-core-utils:published-result');
    should(await actual.serial()).deepEqual(['ok', 'bun-core-utils:published-result']);
  });

  it.each([
    {
      namespace: '!!!',
      key: 'key',
      field: 'namespace' as const,
      message: 'namespace must not be empty',
    },
    {
      namespace: 'namespace',
      key: '!!!',
      field: 'key' as const,
      message: 'key must not be empty',
    },
  ])('should return a published Result Err when $field is empty', async fixture => {
    // Arrange

    // Act
    const actual = namespacedKey(fixture.namespace, fixture.key);
    const error = await actual.unwrapErr();

    // Assert
    should(await actual.isErr()).equal(true);
    should(error).be.instanceof(NamespacedKeyValidationError);
    should(error.field).equal(fixture.field);
    should(error.reason).equal('must not be empty');
    should(error.message).equal(fixture.message);
    should(error.name).equal('NamespacedKeyValidationError');
  });
});
